File: src/main/java/com/pauljoda/modularsystems/core/network/OpenContainerPacket.java
package com.pauljoda.modularsystems.core.network;

import com.teambr.bookshelf.Bookshelf;
import cpw.mods.fml.common.network.simpleimpl.IMessage;
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
import cpw.mods.fml.common.network.simpleimpl.MessageContext;
import io.netty.buffer.ByteBuf;

public class OpenContainerPacket implements IMessageHandler<OpenContainerPacket.UpdateMessage, IMessage> {

    @Override
    public IMessage onMessage(OpenContainerPacket.UpdateMessage message, MessageContext ctx) {
        if (ctx.side.isServer()) {
            ctx.getServerHandler().playerEntity.openGui(Bookshelf.instance, 0,
                    ctx.getServerHandler().playerEntity.getEntityWorld(),
                    message.x, message.y, message.z);
        }
        return null;
    }

    public static class UpdateMessage implements IMessage {

        private int x, y, z;

        @SuppressWarnings("unused")
        public UpdateMessage() {}

        public UpdateMessage(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        @Override
        public void fromBytes(ByteBuf buf) {
            x = buf.readInt();
            y = buf.readInt();
            z = buf.readInt();
        }

        @Override
        public void toBytes(ByteBuf buf) {
            buf.writeInt(x);
            buf.writeInt(y);
            buf.writeInt(z);
        }
    }
}

File: src/main/scala/com/teambr/modularsystems/core/network/OpenContainerPacket.scala
package com.teambr.modularsystems.core.network

import com.teambr.bookshelf.Bookshelf
import io.netty.buffer.ByteBuf
import net.minecraft.util.BlockPos
import net.minecraftforge.fml.common.network.simpleimpl.{MessageContext, IMessage, IMessageHandler}

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 08, 2015
 */
class OpenContainerPacket extends IMessage with IMessageHandler[OpenContainerPacket, IMessage] {

    var blockPosition = new BlockPos(0, 0, 0)

    def this(blockPos: BlockPos) {
        this()
        blockPosition = blockPos
    }

    override def fromBytes(buffer : ByteBuf): Unit = {
        blockPosition = BlockPos.fromLong(buffer.readLong())
    }

    override def toBytes(buffer : ByteBuf): Unit = {
        buffer.writeLong(blockPosition.toLong)
    }

    def onMessage(message: OpenContainerPacket, ctx: MessageContext): IMessage = {
        if (ctx.side.isServer) {
            val pos = message.blockPosition
            ctx.getServerHandler.playerEntity.openGui(Bookshelf, 0, ctx.getServerHandler.playerEntity.getEntityWorld,
                pos.getX, pos.getY, pos.getZ)
        }
        null
    }
}
